### ISSUE TITLE
Try improving NER model performance by using publicly available NER datasets  

### ISSUE THREAD

**Issue by FrancescoCasalegno** [BlueBrain/Search#608]
## Context
- In #605 we tried to improve our NER model by leveraging the NER-annotated sentences generated from the Ontology.
- However, that didn't work because the quality of the annotations was too poor.
- Instead, we can think of using any of the publicly available datasets for NER with biological entity types.
- For instance (but not limited to) we can look to the corpora used by [SciSpaCy](https://allenai.github.io/scispacy/) to train their models: CRAFT, JNLPBA, BC5CDR, BIONLP13CG. 

## Actions
- [ ] Find publicly available annotated NER datasets that cover some of the entity types we also want.  
- [ ] Think about how to handle the fact that some of those dataests may not have annotations for all our entity types, and at the same time they may have annotations for entity types we do not care about.
    - use label `-100` to mask tokens so the `torch` loss function ignores them?
    - train one NER model per entity type? but then how to resolve conflicts?
- [ ] See if learning curves show higher performance (e.g. better intercept + same slope) than what we got in #601 and #602. To have comparable results, we could do the following (?):
     - train on 1/8 of our data + all external datasets
     - train on 2/8 of our data + all external datasets
     - train on 4/8 of our data + all external datasets
     - train on 8/8 of our data + all external datasets

**Comment by EmilieDel** [comment]
`bionlp13cg` has 16 entity types. 
```AMINO_ACID, ANATOMICAL_SYSTEM, CANCER, CELL, CELLULAR_COMPONENT, DEVELOPING_ANATOMICAL_STRUCTURE, GENE_OR_GENE_PRODUCT, IMMATERIAL_ANATOMICAL_ENTITY, MULTI-TISSUE_STRUCTURE, ORGAN, ORGANISM, ORGANISM_SUBDIVISION, ORGANISM_SUBSTANCE, PATHOLOGICAL_FORMATION, SIMPLE_CHEMICAL, TISSUE```

After trying a simple 1:1 correspondence between our entity types and the entity types of the model. 
| Our entity type | Model entity type | 
| -- | -- |
| GENE | GENE_OR_GENE_PRODUCT |
| CELL_TYPE | CELL | 
| BRAIN_REGION | ANATOMICAL_SYSTEM |
| CELL_COMPARTMENT | CELLULAR_COMPONENT | 
| ORGANISM | ORGANISM | 

Here are the results we obtain without any fine tuning:
|        |   precision |   recall  | f1-score  | support |
| -- | -- | -- | -- | -- |
| BRAIN_REGION |      0.23   |    0.18  |     0.20   |     345 | 
| CELL_COMPARTMENT |      0.17 |      0.45   |    0.25   |     177 | 
|  CELL_TYPE  |     0.28    |   0.48   |    0.35    |    677 | 
|       GENE    |   0.55  |     0.67  |     0.60  |     1469 | 
|   ORGANISM    |   0.28    |   0.46  |     0.34  |      279 |

**Comment by EmilieDel** [comment]
Discussed during 26-07 meeting
TO DOs: 
- [ ] k-fold on the internal dataset and compute means of experiments 
- [ ] Train with the internal dataset and then "fine-tune" with the external datasets
- [ ] (Less important) Training phase 1 with M2 approach

**Comment by jankrepl** [comment]
K-fold cross-validation with 5 folds


## F1 score
#### Test
<img width="682" alt="Screenshot 2022-08-02 at 09 38 32" src="https://user-images.githubusercontent.com/18519371/182319322-9f5601a3-8671-4f62-8a43-3e8488d52799.png">

#### Train
<img width="682" alt="Screenshot 2022-08-02 at 09 38 51" src="https://user-images.githubusercontent.com/18519371/182319581-4c0918f4-0bd9-4599-ad25-36d20c06a5a2.png">

**Comment by FrancescoCasalegno** [comment]
## Planning 2022-08-02

- [ ] Look at results of train + eval (k-fold cross-validation) after #607 fixes the annotations in the "ground truth"
- [ ] Try to "pre-train" on the external (partially annotated) NER dataset and then "fine-tune" on the internally (fully annotated) NER datasets

**Comment by jankrepl** [comment]
K-fold cross-validation with 5 folds using the **original** annotations

* `internal` - only trained on fully annotated data
* `external_simul` - fully annotated data + external data were concatenated and the network was trained on this dataset
* `external_seq`- we first trained in external data and then trained on fully annotated internal data (sequential logic)


## F1 score
#### Test
![Screenshot 2022-08-12 at 11 19 05](https://user-images.githubusercontent.com/18519371/184325381-3c5daf73-8436-48e0-9b7b-277c3b0a1c28.png)



#### Train
![Screenshot 2022-08-12 at 11 23 05](https://user-images.githubusercontent.com/18519371/184325439-d7f70428-f388-4f6e-b246-85d6cf2f0114.png)

**Comment by jankrepl** [comment]
K-fold cross-validation with 5 folds using the corrected annotation. The rest is the same as above

#F1-score
## Train
![Screenshot 2022-08-15 at 13 35 51](https://user-images.githubusercontent.com/18519371/184628324-8e18e093-9b62-455e-a11c-6cb0272dcfb0.png)


## Test
![Screenshot 2022-08-15 at 13 35 43](https://user-images.githubusercontent.com/18519371/184628337-4f7b10cd-9146-44d3-b791-bc941888a3f3.png)

**Comment by FrancescoCasalegno** [comment]
## Update 2022-08-16

- Based on the results shown in https://github.com/BlueBrain/Search/issues/608#issuecomment-1212909088, it seems that merging the partially annotated NER samples with the fully annotated ones from BBP gives bad results. Possibly, this is because the partially annotated ones outnumber the fully annotated, high quality ones.
- Based on the results shown in https://github.com/BlueBrain/Search/issues/608#issuecomment-1214915372, pre-training on the partially annotated NER samples does not decrease but neither it significantly improves the accuracy of the final NER model.

**Decision**
- For the time being, it does not seem like we can leverage (partially annotated) publicly available NER datasets to improve the performance of our NER models.
